### PR TITLE
Add dockerfile which is based on Ubuntu 22.04

### PR DESCRIPTION
**Pull request by andr-sokolov**
Add dockerfile which is based on Ubuntu 22.04

There is no sudo in the ubuntu:22.04 image, so sudo is moved from
README.ubuntu.bash to README.linux.md. Add packages, which are required to build
and run GPDB, to README.ubuntu.bash. The new Dockerfile.ubuntu file is similar
to arenadata/Dockerfile. The first RUN setups the environment to run
compile_gpdb.bash and contains all changes to the ubuntu:22.04 image, which can
be made without the source code. So this layer will not be rebuilt for every
commit.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/76568

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/76657

**Comment by BenderArenadata** [comment]
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1718709

**Comment by andr-sokolov** [comment]
bender build

**Comment by RekGRpth** [comment]
> The new Dockerfile.ubuntu file is similar to
> arenadata/Dockerfile.

Why are some steps in a different order?

File: arenadata/Dockerfile.ubuntu [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ARG sigar=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/sigar_1.6.5-3304%2Bgite8961a6_all.deb
+ARG sigar_headers=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/sigar-headers_1.6.5-3304%2Bgite8961a6_all.deb
+ARG libxerces=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/libxerces-c31_3.1.2-3304_amd64.deb
+ARG libxerces_dev=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/libxerces-c31-dev_3.1.2-3304_amd64.deb
+
+WORKDIR /home/gpadmin
+
+COPY README.ubuntu.bash ./
+RUN ./README.ubuntu.bash && \
+    rm README.ubuntu.bash && \
```
Comment (RekGRpth) [inline arenadata/Dockerfile.ubuntu]: centos's docker file does not contain these steps

**Comment by RekGRpth** [comment]
```bash
export GP_MAJOR=6
docker build \
    --build-arg GP_MAJOR=$GP_MAJOR \
    --file "arenadata/Dockerfile.ubuntu" \
    --pull \
    --tag "gpdb${GP_MAJOR}u" \
    . 2>&1 | tee "build${GP_MAJOR}u.log"
...
#13 96.88 /usr/bin/ld: transform.o: relocation R_X86_64_32S against `.rodata' can not be used when making a PIE object; recompile with -fPIE
#13 96.89 /usr/bin/ld: failed to set dynamic section sizes: bad value
#13 96.89 collect2: error: ld returned 1 exit status
#13 96.89 make[4]: *** [Makefile:39: gpfdist] Error 1
#13 96.89 make[3]: *** [Makefile:31: install-gpfdist-recurse] Error 2
#13 96.89 make[3]: *** Waiting for unfinished jobs....
#13 97.31 /usr/bin/ld: pg_backup_tar.o: in function `newTempFile':
#13 97.31 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
#13 98.50 /usr/bin/ld: pg_backup_tar.o: in function `newTempFile':
#13 98.50 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
#13 98.55 make[2]: *** [Makefile:34: install-bin-recurse] Error 2
#13 98.55 make[1]: *** [GNUmakefile:12: install-src-recurse] Error 2
#13 98.55 make: *** [Makefile:318: dist] Error 2
#13 ERROR: process "/bin/sh -c bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash" did not complete successfully: exit code: 2
------
 > [build 3/3] RUN bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash:
96.89 make[4]: *** [Makefile:39: gpfdist] Error 1
96.89 make[3]: *** [Makefile:31: install-gpfdist-recurse] Error 2
96.89 make[3]: *** Waiting for unfinished jobs....
pg_backup_tar.o: in function `newTempFile':
97.31 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
pg_backup_tar.o: in function `newTempFile':
98.50 /home/gpadmin/gpdb_src/src/bin/pg_dump/pg_backup_tar.c:506: warning: the use of `tempnam' is dangerous, better use `mkstemp'
98.55 make[2]: *** [Makefile:34: install-bin-recurse] Error 2
98.55 make[1]: *** [GNUmakefile:12: install-src-recurse] Error 2
98.55 make: *** [Makefile:318: dist] Error 2
------
Dockerfile.ubuntu:29
--------------------
  27 |
  28 |     # Compile with running mocking tests
  29 | >>> RUN bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash
  30 |
  31 |     FROM base as code
--------------------
ERROR: failed to solve: process "/bin/sh -c bash /home/gpadmin/gpdb_src/concourse/scripts/compile_gpdb.bash" did not complete successfully: exit code: 2
```

File: arenadata/Dockerfile.ubuntu [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+ARG sigar=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/sigar_1.6.5-3304%2Bgite8961a6_all.deb
+ARG sigar_headers=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/sigar-headers_1.6.5-3304%2Bgite8961a6_all.deb
+ARG libxerces=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/libxerces-c31_3.1.2-3304_amd64.deb
+ARG libxerces_dev=https://downloads.adsw.io/ADB/6.27.1_arenadata56/ubuntu/22.04/community/x86_64/packages/libxerces-c31-dev_3.1.2-3304_amd64.deb
+
+WORKDIR /home/gpadmin
+
+COPY README.ubuntu.bash ./
+RUN ./README.ubuntu.bash && \
```
Comment (RekGRpth) [inline arenadata/Dockerfile.ubuntu]: try use modern syntax with set -eux and ;

File: arenadata/Dockerfile.ubuntu [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    rm README.ubuntu.bash && \
+    wget $sigar $sigar_headers $libxerces $libxerces_dev && \
+    apt install ./*.deb && \
```
Comment (RekGRpth) [inline arenadata/Dockerfile.ubuntu]: use -y flag with install

File: arenadata/Dockerfile.ubuntu (hunk 1 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: The current approach is better than approach used in centos's docker file, because we keep package list to install in one place

File: arenadata/Dockerfile.ubuntu (hunk 2 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: done

File: arenadata/Dockerfile.ubuntu (hunk 3 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: done

**Comment by andr-sokolov** [comment]
> > The new Dockerfile.ubuntu file is similar to
> > arenadata/Dockerfile.
> 
> Why are some steps in a different order?

I've updated the PR description and moved WORDIR to the same place as in the dockerfile for CentOS

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/76851

**Review by RekGRpth — DISMISSED** [review]

File: arenadata/Dockerfile.ubuntu [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+COPY --from=code /home/gpadmin/gpdb_src gpdb_src
+COPY --from=build /home/gpadmin/bin_gpdb /home/gpadmin/bin_gpdb
+
+# Install entab and pg_bsd_indent used by pgindent utility.
+# This should be done using gpdb sources.
+RUN set -eux; \
+    make -C gpdb_src/src/tools/entab install clean; \
+    wget https://ftp.postgresql.org/pub/dev/pg_bsd_indent-1.3.tar.gz -O - | tar -xzf -; \
+    make install -C pg_bsd_indent; \
+    rm -r pg_bsd_indent; \
```
Comment (KnightMurloc) [inline arenadata/Dockerfile.ubuntu]: Why is this necessary? We don't do this in a container for centos.

File: arenadata/Dockerfile.ubuntu [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# libipc-run-perl and rsync are needed to run regression tests
+# iproute2 is needed to run GPDB
```
Comment (KnightMurloc) [inline arenadata/Dockerfile.ubuntu]: if this package is needed for gpdb to work, why not add it to `README.ubuntu.bash`?

File: arenadata/Dockerfile.ubuntu [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    apt install -y libipc-run-perl rsync iproute2; \
+# The en_US.UTF-8 locale is needed to run GPDB
+    locale-gen en_US.UTF-8; \
+# To run sshd directly, but not using `/etc/init.d/ssh start`
+    mkdir /run/sshd; \
+# Alter precedence in favor of IPv4 during resolving
+    echo 'precedence ::ffff:0:0/96  100' >> /etc/gai.conf
```
Comment (KnightMurloc) [inline arenadata/Dockerfile.ubuntu]: can we do this before compiling gpdb so that we don't have to do it every time we rebuild?

**Review by KnightMurloc — COMMENTED** [review]
There are not enough packages to run the behave tests. at least the behave package itself. There also a java package missing, which I understand is needed for pxf tests, or will pxf be tested in another container? There are also no more java related steps https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L80-L81.

File: arenadata/Dockerfile.ubuntu (hunk 4 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L88C1-L91C25

**Comment by andr-sokolov** [comment]
> There are not enough packages to run the behave tests. at least the behave package itself. There also a java package missing, which I understand is needed for pxf tests, or will pxf be tested in another container? There are also no more java related steps
> 
> https://github.com/arenadata/gpdb/blob/6ad72daf603ca7b061863e6e7354d45e44e24e89/arenadata/Dockerfile#L80-L81
> .

This image version purpose is to successfully execute compile_gpdb.bash and run compiled GPDB. Compatibility with all the tests will be checked and implemented on the next steps

File: arenadata/Dockerfile.ubuntu (hunk 6 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: done

File: arenadata/Dockerfile.ubuntu (hunk 5 above)
Comment (andr-sokolov) [reply arenadata/Dockerfile.ubuntu]: done

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/76944

**Comment by BenderArenadata** [comment]
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1736023

**Review by KnightMurloc — APPROVED** [review]

**Review by RekGRpth — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/76989
